access/ms-teams/msapi/bot_framework_client.go

```diff
@@ -0,0 +1,135 @@
+package msapi
+
+import (
+	"context"
+	"net/http"
+
+	"github.com/gravitational/trace"
+)
+
+const (
+	botFrameworkAuthScope     = "https://api.botframework.com/.default"
+	botFrameworkBaseURL       = "https://smba.trafficmanager.net"
+	botFrameworkDefaultRegion = "uk"
+	botFrameworkVersion       = "v3"
+)
+
+// BotFrameworkClient represents client to MS Graph API
+type BotFrameworkClient struct {
+	Client
+}
+
+// ChatMember represents member of a chat
+type ChatMember struct {
+	ID                string `json:"id"`
+	ObjectID          string `json:"objectId"`
+	Name              string `json:"name"`
+	Email             string `json:"email"`
+	UserPrincipalName string `json:"userPrincipalName"`
+	TenantID          string `json:"tenantId"`
+	UserRole          string `json:"userRole"`
+}
+
+// PostActivityResponse represents json response with a single id field
+type PostActivityResponse struct {
+	ID string `json:"id"`
+}
+
+// botError represents MS Graph error
+type botError struct {
+	E struct {
+		Code    string `json:"code"`
+		Message string `json:"message"`
+	} `json:"error"`
+}
+
+// Error returns error string
+func (e *botError) Error() string {
+	return e.E.Code + " " + e.E.Message
+}
+
+// NewBotFrameworkClient creates MS Graph API client
+func NewBotFrameworkClient(config Config) *BotFrameworkClient {
+	region := config.Region
+	if region == "" {
+		region = botFrameworkDefaultRegion
+	}
+
+	baseURL := config.url.botFrameworkBaseURL
+	if baseURL == "" {
+		baseURL = botFrameworkBaseURL
+	}
+
+	baseURL = baseURL + "/" + region + "/" + botFrameworkVersion
+
+	return &BotFrameworkClient{
+		Client: Client{
+			token:   tokenWithTTL{scope: botFrameworkAuthScope, baseURL: config.url.tokenBaseURL},
+			baseURL: baseURL,
+			config:  config,
+		},
+	}
+}
+
+// GetChatMembers returns members of a chat
+func (c *BotFrameworkClient) GetChatMembers(ctx context.Context, chatID string) ([]ChatMember, error) {
+	var m []ChatMember
+
+	request := request{
+		Method:   http.MethodGet,
+		Path:     "conversations/" + chatID + "/members",
+		Response: &m,
+		Err:      &botError{},
+	}
+
+	err := c.request(ctx, request)
+	if err != nil {
+		return nil, trace.Wrap(err)
+	}
+
+	return m, nil
+}
+
+// PostAdaptiveCardActivity sends an activity to the chat, content is AdaptiveCard
+func (c *BotFrameworkClient) PostAdaptiveCardActivity(ctx context.Context, botID, chatID, card, updateID string) (string, error) {
+	body := `
```

Comment:
Can we convert this into a struct and call jsonMarshal?
It's an extra call, but I would rather let go build the json automatically